rewrite-java/src/main/java/org/openrewrite/java/cleanup/MakePrivateOrFinalMethodStaticIfPossible.java

```diff
@@ -0,0 +1,234 @@
+        if (classDeclaration == null) {
+            return false;
+        }
+        if (TypeUtils.isAssignableTo("java.io.Serializable", classDeclaration.getType())) {
+            // this is potentially over-broad if the methods are overloaded, but that seems uncommon
+            switch (method.getSimpleName()) {
+                case "readObject":
+                case "readObjectNoData":
+                case "readResolve":
+                case "writeObject":
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    private static boolean doesMethodMaybeReferenceThis(J.MethodDeclaration method, Cursor cursor) {
+        JavaIsoVisitor<AtomicBoolean> visitor = new JavaIsoVisitor<AtomicBoolean>() {
+
+            @Override
+            public J.NewClass visitNewClass(J.NewClass newClass, AtomicBoolean maybeHasThis) {
+                if (isImplicitEnclosingThisNewClass(newClass)) {
+                    maybeHasThis.set(true);
+                } else if (newClass.getBody() != null) {
+                    // inside the anonymous class body, it's harder to determine which scope
+                    // an implied `this` refers to, so for now we just bail
+                    maybeHasThis.set(true);
+                }
+                return super.visitNewClass(newClass, maybeHasThis);
+            }
+
+            @Override
+            public J.Identifier visitIdentifier(J.Identifier identifier, AtomicBoolean maybeHasThis) {
+                if (isExplicitThis(identifier) || isImplicitThisFieldAccess(identifier, getCursor()) || isImplicitThisMethodInvocation(identifier, getCursor())) {
+                    maybeHasThis.set(true);
+                }
+                return super.visitIdentifier(identifier, maybeHasThis);
+            }
+        };
+
+        AtomicBoolean hasThis = new AtomicBoolean(false);
+        visitor.visit(method.getBody(), hasThis);
+        return hasThis.get();
+    }
+
+    private static boolean isExplicitThis(J.Identifier identifier) {
+        return identifier.getSimpleName().equals("this");
+    }
+
+    private static boolean isImplicitThisMethodInvocation(J.Identifier identifier, Cursor cursor) {
+        Object parent = cursor.getParentTreeCursor().getValue();
+
+        // not part of a method invocation
+        if (!(parent instanceof J.MethodInvocation)) {
+            return false;
+        }
+
+        // method invocation has a target
+        // e.g. `x.identifier()`
+        if (((J.MethodInvocation) parent).getSelect() != null) {
+            return false;
+        }
+
+        // identifier is not the name of the method invocation
+        if (((J.MethodInvocation) parent).getName().getId() != identifier.getId()) {
+            return false;
+        }
+
+        // not an instance method
+        JavaType.@Nullable Method methodType = ((J.MethodInvocation) parent).getMethodType();
```

Comment:
I really appreciate how the annotation is specific to the inner class here. Since Java now permits annotations in more places than it once did, a recipe on moving annotations to the most granular level would be cool.